CIPs/cip-0032.md

```diff
@@ -0,0 +1,117 @@
+
+### Nothing at stake problem
+
+Unlike other slashing contracts which use `SlasherUtil.performSlashing()` to both slash _and_ remove the validator, this `AttestationSlasher` would directly call `LockedGold.slash()`, only slashing stake and not removing the validator. This decreases the potential risk of having multiple validators removed simultaneously due to an overly aggressive `AttestationSlasher` policy. If the validator is slashed it may be that their locked stake drops below the `validatorLockedGoldRequirements` (currently 10K CELO). By dropping below `validatorLockedGoldRequirements` alone, the validator is not removed from the validator set but they will cease to earn epoch rewards. If they continue to get slashed by the `AttestationSlasher`, it may be that they drop to a near-zero stake while still remaining to be a validator. This is a dangerous situation, since this validator is performing an important role for the network but has nothing at stake. To resolve this tension, we'll introduce a new `MinStakeSlasher` which will force a validator to be de-affiliated if their balance drops below the `penalty + reward` threshold of any of the registered slasher contracts. Once a validator has been de-affiliated, they'll need to return their stake amount to the `validatorLockedGoldRequirements` requirement before being able to affiliate again.
+
+### Defining parameters
+
+#### AttestationSlasher
+
+- `B` This will be chosen based on the minimal period that it is likely for a validator to have recieved an attestation request. We'll start conservative and look at a period where every validator should have received 10 attestations. In the month of November 2020, there was an average of 94 attestations requested per day (more in December). Therefore, over 11 days, an average of 1034 attesations would be requested, providing an expected value of ~10 attestations for each of the 100 issuers. Since we produce a block every 5 seconds, 11 days is equivalent to 190,080 blocks.
+- `P` Similarly we'll start conservatively with the amount of stake slashed. At the time of writing this, only two active validators have an excess stake of less than 100 CELO (85 CELO each). If we set a penalty of 40 CELO, then these validators can be slashed twice before incurring any loss in epoch rewards.
+- `R` The reward for slashing should cover both the gas costs as well as the infrastructure costs to monitor for offenses.
+
+#### MinStakeSlasher
+
+- `P` The penalty is zero since the validator has already been slashed to get to this point.
+- `R` Similarly, the reward for slashing should cover both the gas costs as well as the infrastructure costs to monitor for offenses.
```

Comment:
I'm not super familiar with where the funds for slashing rewards come from. Do they get distributed with the general epoch rewards or does a portion of the penalty go to the slashing? If the later, how do we fund the reward if the penalty becomes 0?

Reply:
The rewards come from the staked funds that are being slashed. The reward is first given, then the remaining amount is slashed. You're correct that the penalty should actually be equal to the reward in this case. I was originally thinking about the penalty being the addition to the reward but it's clearer if we assume it to be the total amount.